File: modules/wem-api/src/main/java/com/enonic/wem/api/data/Property.java
package com.enonic.wem.api.data;


import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.Objects;

import org.apache.commons.lang.StringUtils;

import com.google.common.base.Preconditions;

import com.enonic.wem.api.util.BinaryReference;
import com.enonic.wem.api.util.GeoPoint;
import com.enonic.wem.api.util.Link;
import com.enonic.wem.api.util.Reference;

public final class Property
{
    private final PropertySet parent;

    private final String name;

    private final int index;

    private Value value;

    private PropertyId id;

    Property( final String name, final int index, final Value value, final PropertyId id, final PropertySet parent )
    {
        checkName( name );
        this.name = name;
        this.index = index;
        this.value = value;
        this.id = id;
        this.parent = parent;
        setPropertyOnPropertySetValue( value, parent );
    }

    private void setPropertyOnPropertySetValue( final Value value, final PropertySet parent )
    {
        if ( value.isSet() && !value.isNull() )
        {
            value.asData().setProperty( this );

            checkPropertySetValueIsNotRootPropertySet( value, parent );
        }
    }

    private void checkPropertySetValueIsNotRootPropertySet( final Value value, final PropertySet parent )
    {
        if ( parent.getTree() != null )
        {
            if ( value.asData() == parent.getTree().getRoot() )
            {
                throw new IllegalArgumentException( "Given PropertySet is already the root PropertySet of the PropertyTree" );
            }
        }
    }

    void setId( final PropertyId id )
    {
        Preconditions.checkState( this.id == null, "id already set" );
        this.id = id;
    }

    void detach()
    {
        if ( value.getType().equals( ValueTypes.PROPERTY_SET ) && !value.isNull() )
        {
            value.asData().detach();
        }
    }

    public void setValue( final Value value )
    {
        this.value = value;
        this.setPropertyOnPropertySetValue( value, this.parent );
    }

    public PropertySet getParent()
    {
        return parent;
    }

    PropertyId getId()
    {
        return id;
    }

    public String getName()
    {
        return name;
    }

    public int getIndex()
    {
        return index;
    }

    boolean hasParentProperty()
    {
        return getParentProperty() != null;
    }

    Property getParentProperty()
    {
        return this.parent.getProperty();
    }

    public PropertyPath getPath()
    {
        if ( hasParentProperty() )
        {
            return PropertyPath.from( this.parent.getProperty().getPath(), PropertyPath.Element.from( this.name, this.index ) );
        }
        else
        {
            return PropertyPath.from( PropertyPath.Element.from( this.name, this.index ) );
        }
    }

    public ValueType getType()
    {
        return value.getType();
    }

    public Value getValue()
    {
        return value;
    }

    public boolean hasNullValue()
    {
        return value.isNull();
    }

    public boolean hasNotNullValue()
    {
        return !value.isNull();
    }

    public PropertySet getSet()
    {
        return value.asData();
    }

    public Object getObject()
    {
        return value.getObject();
    }

    public String getString()
    {
        return value.asString();
    }

    public Boolean getBoolean()
    {
        return value.asBoolean();
    }

    public Long getLong()
    {
        return value.asLong();
    }

    public Double getDouble()
    {
        return value.asDouble();
    }

    public GeoPoint getGeoPoint()
    {
        return value.asGeoPoint();
    }

    public Reference getReference()
    {
        return value.asReference();
    }

    public BinaryReference getBinaryReference()
    {
        return value.asBinaryReference();
    }

    public Link getLink()
    {
        return value.asLink();
    }

    public LocalDate getLocalDate()
    {
        return value.asLocalDate();
    }

    public LocalDateTime getLocalDateTime()
    {
        return value.asLocalDateTime();
    }

    public LocalTime getLocalTime()
    {
        return value.asLocalTime();
    }

    public Instant getInstant()
    {
        return value.asInstant();
    }

    public static void checkName( final String name )
    {
        if ( name == null )
        {
            throw new NullPointerException( "Property name cannot be null" );
        }
        if ( StringUtils.isBlank( name ) )
        {
            throw new IllegalArgumentException( "Property name cannot be blank" );
        }
        if ( name.contains( "." ) )
        {
            throw new IllegalArgumentException( "Property name cannot contain ." );
        }
        if ( name.contains( "[" ) || name.contains( "]" ) )
        {
            throw new IllegalArgumentException( "Property name cannot contain [ or ]" );
        }
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( !( o instanceof Property ) )
        {
            return false;
        }

        final Property property = (Property) o;

        if ( !id.equals( property.id ) )
        {
            return false;
        }
        if ( !name.equals( property.name ) )
        {
            return false;
        }
        if ( index != property.index )
        {
            return false;
        }
        if ( !value.equals( property.value ) )
        {
            return false;
        }

        return Objects.equals( index, property.index ) && Objects.equals( name, property.name ) && Objects.equals( value, property.value );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( name, index, value );
    }

    @Override
    public String toString()
    {
        final StringBuilder s = new StringBuilder();
        s.append( name ).append( ": " ).append( value );
        return s.toString();
    }

    public int countAncestors()
    {
        return getPath().elementCount() - 1;
    }

    /**
     * Copies this Property and adds it to the given PropertySet.
     */
    public Property copyTo( final PropertySet destination )
    {
        final Value copiedValue = value.copy( destination.getTree() );
        final Property property = new Property( name, index, copiedValue, id, destination );
        destination.add( property );
        return property;
    }
}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/data/Property.ts
module api.data {

    export class Property implements api.Equitable {

        private parent: PropertySet;

        private name: string;

        private index: number;

        private value: Value;

        private id: PropertyId;

        private propertyChangedListeners: {(event: PropertyChangedEvent):void}[];

        constructor(builder: PropertyBuilder) {
            api.util.assertNotNull(builder.parent, "parent of a Property cannot be null");
            api.util.assertNotNull(builder.name, "name of a Property cannot be null");
            api.util.assertNotNull(builder.index, "index of a Property cannot be null");
            api.util.assertNotNull(builder.value, "value of a Property cannot be null");
            api.util.assertNotNull(builder.id, "id of a Property cannot be null");

            this.parent = builder.parent;
            this.name = builder.name;
            this.index = builder.index;
            this.value = builder.value;
            this.id = builder.id;

            if (this.value.getType().equals(ValueTypes.DATA) && this.value.isNotNull()) {
                this.value.getPropertySet().setContainerProperty(this);
            }

            this.propertyChangedListeners = [];
        }

        setIndex(index: number) {
            this.index = index;
            // TODO: fire event?
        }

        setValue(value: Value) {
            api.util.assertNotNull(value, "value of a Property cannot be null");
            var oldValue = this.value;
            this.value = value;
            if (this.value.getType().equals(ValueTypes.DATA) && this.value.isNotNull()) {
                this.value.getPropertySet().setContainerProperty(this);
            }
            if (!value.equals(oldValue)) {
                this.notifyPropertyChangedEvent(new PropertyChangedEvent(PropertyChangedEventType.CHANGED, this.getPath(), value));
            }
        }

        getId(): PropertyId {
            return this.id;
        }

        getParent(): PropertySet {
            return this.parent;
        }

        hasParentProperty(): boolean {
            return !!this.getParentProperty();
        }

        getParentProperty(): Property {
            return this.parent.getProperty();
        }

        getPath(): PropertyPath {
            if (this.hasParentProperty()) {
                return PropertyPath.fromParent(this.getParentProperty().getPath(), new PropertyPathElement(this.name, this.index))
            }
            else {
                return PropertyPath.fromPathElement(new PropertyPathElement(this.name, this.index));
            }
        }

        getName(): string {
            return this.name;
        }

        getIndex(): number {
            return this.index;
        }

        getType(): ValueType {
            return this.value.getType();
        }

        getValue(): Value {
            return this.value;
        }

        hasNullValue(): boolean {
            return this.value.isNull();
        }

        hasNonNullValue(): boolean {
            return !this.value.isNull();
        }

        getString(): string {
            return this.value.getString();
        }

        getBoolean(): boolean {
            return this.value.getBoolean();
        }

        getDate(): Date {
            return this.value.getDate();
        }

        getSet(): PropertySet {
            return this.value.getPropertySet();
        }

        getGeoPoint(): api.util.GeoPoint {
            return this.value.getGeoPoint();
        }

        getLocalTime(): api.util.LocalTime {
            return this.value.getLocalTime();
        }

        getContentId(): api.content.ContentId {
            return this.value.getContentId();
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, Property)) {
                return false;
            }

            var other = <Property>o;

            if (!api.ObjectHelper.equals(this.id, other.id)) {
                return false;
            }

            if (!api.ObjectHelper.stringEquals(this.name, other.name)) {
                return false;
            }

            if (!api.ObjectHelper.numberEquals(this.index, other.index)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.value, other.value)) {
                return false;
            }

            return true;
        }

        copy(destinationPropertySet: PropertySet, generateNewPropertyId: boolean = false) {
            var copy = Property.create().
                setId(generateNewPropertyId ? destinationPropertySet.getTree().getNextId() : this.id).
                setName(this.name).
                setValue(this.value).
                setIndex(this.index).
                setParent(destinationPropertySet).
                build();
            return copy;
        }

        prettyPrint(indent?: string) {
            var thisIndent = indent ? indent : "";
            var idAsString = this.getId().toString();
            var valueAsString = this.getValue().getString();
            console.log(thisIndent + idAsString + ": " + valueAsString);
        }

        onPropertyChanged(listener: {(event: PropertyChangedEvent): void;}) {
            this.propertyChangedListeners.push(listener);
        }

        unPropertyChanged(listener: {(event: PropertyChangedEvent): void;}) {
            this.propertyChangedListeners =
            this.propertyChangedListeners.filter((curr) => (curr != listener));
        }

        private notifyPropertyChangedEvent(event: PropertyChangedEvent) {
            this.propertyChangedListeners.forEach((listener) => listener(event));
        }

        public static checkName(name: string) {
            if (name == null) {
                throw new Error("Property name cannot be null");
            }
            if (api.util.StringHelper.isBlank(name)) {
                throw new Error("Property name cannot be blank");
            }
            if (name.indexOf(".") >= 0) {
                throw new Error("Property name cannot contain .");
            }
            if (name.indexOf("[") >= 0 || name.indexOf("]") >= 0) {
                throw new Error("Property name cannot contain [ or ]");
            }
        }

        public static create(): PropertyBuilder {
            return new PropertyBuilder();
        }
    }

    export class PropertyBuilder {

        parent: PropertySet;

        name: string;

        index: number;

        value: Value;

        id: PropertyId;

        setParent(value: PropertySet): PropertyBuilder {
            this.parent = value;
            return this;
        }

        setName(value: string): PropertyBuilder {
            this.name = value;
            return this;
        }

        setIndex(value: number): PropertyBuilder {
            this.index = value;
            return this;
        }

        setValue(value: Value): PropertyBuilder {
            this.value = value;
            return this;
        }

        setId(value: PropertyId): PropertyBuilder {
            this.id = value;
            return this;
        }

        build(): Property {
            return new Property(this);
        }
    }
}